File: starters/web/src/application/task-service.ts
import { createTask, failure, toggleTask, validateTitle } from '../domain/task';
import type { Result, Task } from '../domain/task';

export interface TaskRepository {
  list(): Promise<Result<readonly Task[]>>;
  add(task: Task): Promise<Result<Task>>;
  // Transform the last observed value. Remote adapters use its version; failures
  // can have an unknown commit outcome and must not trigger automatic retries.
  update(id: string, transform: (current: Task) => Task): Promise<Result<Task>>;
}

export interface TaskService {
  list(): Promise<Result<readonly Task[]>>;
  add(title: unknown): Promise<Result<Task>>;
  toggle(id: string): Promise<Result<Task>>;
}

export function createTaskService(dependencies: Readonly<{
  repository: TaskRepository;
  nextId: () => string;
  now: () => number;
}>): TaskService {
  const { repository, nextId, now } = dependencies;

  async function storage<T>(operation: () => Promise<Result<T>>): Promise<Result<T>> {
    try {
      return await operation();
    } catch {
      // Do not leak adapter exceptions or automatically repeat uncertain effects.
      return failure('STORAGE_UNAVAILABLE');
    }
  }

  return {
    list: () => storage(() => repository.list()),
    async add(input) {
      const title = validateTitle(input);
      if (!title.ok) return title;
      let task: Result<Task>;
      try {
        task = createTask({ id: nextId(), title: title.value, createdAtEpochMs: now() });
      } catch {
        return failure('DEPENDENCY_FAILURE');
      }
      if (!task.ok) return task;
      return storage(() => repository.add(task.value));
    },
    toggle: (id) => storage(() => repository.update(id, toggleTask)),
  };
}

File: starters/kotlin-android/core/src/main/kotlin/org/example/foundation/core/application/TaskService.kt
// SPDX-FileCopyrightText: 2026 Zendrhax LLC
// SPDX-License-Identifier: MPL-2.0
package org.example.foundation.core.application

import org.example.foundation.core.domain.Task
import org.example.foundation.core.domain.TaskError
import org.example.foundation.core.domain.TaskResult
import org.example.foundation.core.domain.createTask
import org.example.foundation.core.domain.toggleTask
import org.example.foundation.core.domain.validateTitle

interface TaskRepository {
    fun list(): TaskResult<List<Task>>
    fun add(task: Task): TaskResult<Task>
    /** Transform the last observed value; remote writes use its version. A failure may have an unknown commit outcome. */
    fun update(id: String, transform: (Task) -> Task): TaskResult<Task>
}

class TaskService(
    private val repository: TaskRepository,
    private val nextId: () -> String,
    private val now: () -> Long,
) {
    fun list(): TaskResult<List<Task>> = storage { repository.list() }

    fun add(title: Any?): TaskResult<Task> {
        val validated = validateTitle(title)
        if (validated is TaskResult.Failure) return validated
        val task = try {
            createTask(nextId(), (validated as TaskResult.Success).value, now())
        } catch (_: Exception) {
            return TaskResult.Failure(TaskError.DEPENDENCY_FAILURE)
        }
        return when (task) {
            is TaskResult.Failure -> task
            is TaskResult.Success -> storage { repository.add(task.value) }
        }
    }

    fun toggle(id: String): TaskResult<Task> = storage { repository.update(id, ::toggleTask) }

    private inline fun <T> storage(operation: () -> TaskResult<T>): TaskResult<T> = try {
        operation()
    } catch (_: Exception) {
        // No automatic retry: an external adapter could have applied an uncertain effect.
        TaskResult.Failure(TaskError.STORAGE_FAILURE)
    }
}
